flytekit/remote/remote.py

```diff
@@ -208,8 +208,14 @@ def _get_pickled_target_dict(
     :param root_entity: The entity to get the pickled target for.
         :return: hashed bytes and the pickled target dictionary.
     """
+    import sys
+
     queue: typing.List[typing.Union[WorkflowBase, PythonTask, CoreNode]] = [root_entity]
-    pickled_target_dict = {}
+    pickled_target_dict = {
+        "metadata": {
```

Comment:
but this shares a namespace with task names right?  so if someone makes a task where `entity.name` is "metadata" then this metadata will be lost and when we go to look for it we'll get a KeyError or something.

I don't think it needs to be here though, this is a pretty standard thing for all tasks.  There's a couple places I think we can fit this information.
* RuntimeMetadata's `flavor` [field](https://github.com/flyteorg/flyte/blob/13b3d82c94fb61047a49aa2093a2b17fb6a7c792/flyteidl/protos/flyteidl/core/tasks.proto#L86).  I don't think this is being used right now.  The name "runtime" is not correct.  It's more like "compile-time", but that's a different problem.
* Something in these [tags](https://github.com/flyteorg/flyte/blob/13b3d82c94fb61047a49aa2093a2b17fb6a7c792/flyteidl/protos/flyteidl/core/tasks.proto#L128).  To my knowledge these tags aren't really being used right now.

I think i have a slight preference for the first one.  what do you think @eapolinario?

Reply:
I am not sure the first approach will be enough though. It is possible that we have to extend it to more fields (like maybe the package version?) in the future, and the first approach will have only 1 string that can store this value.

I think that this metadata is primarily used for interactive mode since only pickling objects will potentially cause the version mismatch problem. When creating a pickled object, the versions should follow with this object. I am thinking of moving the entities dictionary one layer down, so that user could have define function with name `metadata`
```
{
  "metadata": {
    "python_version": "3.12",
    ...
  },
  "entity_dict": {
    "task1": ...
    ...
  }
}
```

Reply:
your call.  I like the idea of adding the `entity_dict` key in general.  I think we should do that anyways.

Your call on the metadata.  The issue with just adding fields is that this risks becoming an undocumented API, so I was thinking we'd just use the existing flavor until we needed more than one field.  If you go the pickled dict route though, just make sure it's well documented/commented.

Reply:
I see your concern. How about creating a dataclass to define the pickled object? In that way, people could know what is inside this object.

Regarding the metadata, the pickling only affects the Python side and doesn't need to be stored in the backend. The main issue this PR addresses is the Python version mismatch between pickled files and remote execution. By sending the Python version inside the pickled file, we can ensure the correctness of the version. Also as far as I researched, I think that `runtimeMetadata` is currently used for flytekit's language and version information rather than the Python version.

Reply:
sounds good!  just make sure using dataclass doesn't break the pickling.